libraries/ESP8266WebServer/src/Parsing-impl.h

```diff
@@ -442,74 +443,64 @@ bool ESP8266WebServerTemplate<ServerType>::_parseForm(ClientType& client, const
-                  goto readfile;
+            int bLen = boundary.length();
+            uint8_t boundBuf[2 + bLen + 1]; // "--" + boundary + null terminator
+            boundBuf[2 + bLen] = '\0';
+            uint8_t argByte;
+            bool first = true;
+            while (1) {
+                //attempt to fill up boundary buffer with length of boundary string
+                int i;
+                for (i = 0; i < 2 + bLen; i++) {
+                    if (!client.connected()) return _parseFormUploadAborted();
+                    argByte = _uploadReadByte(client);
+                    if (argByte == '\r')
+                        break;
+                    boundBuf[i] = argByte;
                 }
-              }
-
-              uint8_t endBuf[boundary.length()];
-              client.readBytes(endBuf, boundary.length());
-
-              if (strstr((const char*)endBuf, boundary.c_str()) != NULL){
-                if(_currentHandler && _currentHandler->canUpload(_currentUri))
-                  _currentHandler->upload(*this, _currentUri, *_currentUpload);
-                _currentUpload->totalSize += _currentUpload->currentSize;
-                _currentUpload->status = UPLOAD_FILE_END;
-                if(_currentHandler && _currentHandler->canUpload(_currentUri))
-                  _currentHandler->upload(*this, _currentUri, *_currentUpload);
-                DBGWS("End File: %s Type: %s Size: %d\n",
-                    _currentUpload->filename.c_str(),
-                    _currentUpload->type.c_str(),
-                    (int)_currentUpload->totalSize);
-                line = client.readStringUntil(0x0D);
-                client.readStringUntil(0x0A);
-                if (line == "--"){
-                  DBGWS("Done Parsing POST\n");
-                  break;
+                if ((strncmp((const char*)boundBuf, "--", 2) == 0) && (strcmp((const char*)(boundBuf + 2), boundary.c_str()) == 0))
+                    break;   //found the boundary, done parsing this file
+                if (first) first = false;   //only add newline characters after the first line
+                else {
+                    _uploadWriteByte('\r');
+                    _uploadWriteByte('\n');
                 }
-                continue;
-              } else {
-                _uploadWriteByte(0x0D);
-                _uploadWriteByte(0x0A);
-                _uploadWriteByte((uint8_t)('-'));
-                _uploadWriteByte((uint8_t)('-'));
-                uint32_t i = 0;
-                while(i < boundary.length()){
-                  _uploadWriteByte(endBuf[i++]);
+                // current line does not contain boundary, upload all bytes in boundary buffer
+                for (int j = 0; j < i; j++)
+                    _uploadWriteByte(boundBuf[j]);
+                // the initial pass (filling up the boundary buffer) did not reach the end of the line. Upload the rest of the line now
+                if (i >= 2 + bLen) {
```

Comment:
I can only ever be <= 2+blen per the above for() loop, so this is really checking `if (i==2+blen)`.  It's not a big deal, but made me do a double-take as I thought there was an overflow condition...

Reply:
I have a habit of using <= or >= instead of == to avoid bugs where a variable gets incremented one too many times and skips the exit case for an infinite loop (for example). Maybe it is a bad habit if it confuses people reading my code!

Reply:
No, you're fine.  It's safer (protects against spurious bitflips, too) and logically correct!